DR2-1358 Create anonymiser lambda code.

MancunianSam opened the pull request:
I've created a project as a workspace and added in a module for common
anonymiser code.
There is a new module for the lambda which uses this code as well as the
existing script code which uses it too.

There are more tests now and a package with common library functions
which are shared between the lambda and the script

techncl commented on testlib/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "testlib"
+version = "0.1.0"
+edition = "2021"
```
Is "2021" correct?

MancunianSam replied:
Yep

techncl commented on anonymise/src/lib.rs:
```diff
@@ -0,0 +1,328 @@
+) -> Result<String, Error> {
+    let docx_file_name: &str = metadata_json_value["parameters"]["TRE"]["payload"]["filename"]
+        .as_str()
+        .ok_or("Filename is missing from the metadata json")
```
Maybe just for clarity "'filename' is missing from the metadata.json" as it might seems as though the metadata.json's file name is missing?

techncl commented on lambda/src/lib.rs:
```diff
@@ -0,0 +1,202 @@
+    upload(&s3_client, &output_tar_path, &output_bucket, &file_name).await?;
+
+    let output_queue = std::env::var("OUTPUT_QUEUE")?;
+    let reference = parameters.reference;
```
Maybe `tre_reference`?

MancunianSam replied:
I think it's ok as is. It's called reference in the json.

techncl commented on lambda/src/lib.rs:
```diff
@@ -0,0 +1,202 @@
+
+    let output_queue = std::env::var("OUTPUT_QUEUE")?;
+    let reference = parameters.reference;
+    let status = parameters.status;
```
Maybe `tre_status`?

techncl reviewed (commented):
Looks good! Just some comments on ways to make things clearer, as well as..

1. The names of the directories are a bit vague
   1. Maybe instead of the directory `anonymise`, it could be called `anonymiser_lib`, `anonymiser_lib`?
   2. Maybe the outermost  `src` directory could be called `anonymiser_script` or `anonymiser_local_script`?

3. Do the libraries have to be called `lib.rs`? Could they be named with the same name as the function?
4. in the cases, when formatting, where you leave the curly braces empty e.g. `format!("/{}/{}", test_input_bucket, test_download_key)`, do you need to? I've seen cases where you add the variable to the curly braces and I think that is more readable

MancunianSam commented:
Thanks for looking at this:
1. ii We can change it to `anonymiser_lib` That is clearer, I've changed it.
    ii I don't think I can change the outer src without a lot of faff so I've created another module called `anonymiser_local_script` That seems to make more sense.
2. Technically they can be renamed but it seems to be the convention to call them that and calling them lib.rs means the library works without any other configuration. As we're still learning all this I think it's best to stick with convention.
3. I forgot you could do that with `format!` I've changed them all.

MancunianSam commented on anonymise/src/lib.rs:
```diff
@@ -0,0 +1,328 @@
+    decompress_file(file, dir_output)?;
+
+    let metadata_input_file_path: &PathBuf = &extracted_output_path.join(PathBuf::from(format!(
+        "TRE-{input_batch_reference}-metadata.json"
```
1. I keep forgetting you can do format! like that. I've changed the one on line 89.
2. Yeah, I've done it as a closure inside the function so I don't have to pass in the `extracted_output_path` which would make it almost as long as what we've got now.

MancunianSam commented on anonymise/src/lib.rs:
```diff
@@ -0,0 +1,328 @@
+        output_batch_reference,
+    )?;
+
+    let _ = fs::remove_dir_all(&extracted_output_path);
```
No, it needs a `?` at the end instead. I blame IntelliJ suggestions.

MancunianSam commented on anonymise/src/lib.rs:
```diff
@@ -0,0 +1,328 @@
+/// * It renames the folder and metadata file from TDR-xxx to TST-xxx.
+/// * It creates a new tar.gz folder in the output directory.
+/// * It deletes the uncompressed folder in the output directory.
+pub fn process_package(dir_output: &PathBuf, file: &PathBuf) -> Result<PathBuf, Error> {
```
1. process_package is ok. We've already got anonymise everywhere 
2. You asked this last time :smile: It's tricky because lots of variables are used throughout the function so you end up with the smaller functions having to return multiple variables which looks messy.

MancunianSam replied:
1. process_package is ok. We've already got anonymise everywhere 
2. You asked this last time :smile: It's tricky because lots of variables are used throughout the function so you end up with the smaller functions having to return multiple variables which looks messy.

MancunianSam commented on lambda/tests/integration_test.rs:
```diff
@@ -0,0 +1,267 @@
+    let send_message_request = sqs_requests.last().unwrap();
+    let body = &send_message_request.body;
+    let sqs_message_string = String::from_utf8(body.to_vec()).unwrap();
+    let expected_string = "{\"QueueUrl\":\"https://example.com\",\"MessageBody\":\"{\\\"parameters\\\":{\\\"status\\\":\\\"ok\\\",\\\"reference\\\":\\\"test-reference\\\",\\\"s3Bucket\\\":\\\"test-output-bucket\\\",\\\"s3Key\\\":\\\"TST-2023.tar.gz\\\"}}\"}";
```
Good shout. I've replaced this one and any other instances where I can see `\"`

MancunianSam commented on anonymise/Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "anonymise"
+version = "0.1.0"
+edition = "2021"
```
Yep, looks like  https://doc.rust-lang.org/edition-guide/rust-2021/index.html

techncl replied:
Ah, I think I misunderstood; I thought it was the edition of our library.

MancunianSam commented on lambda/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+[package]
+name = "lambda"
+version = "0.1.0"
+edition = "2021"
```
As above

MancunianSam commented on lambda/src/lib.rs:
```diff
@@ -0,0 +1,202 @@
+    let output_queue = std::env::var("OUTPUT_QUEUE")?;
+    let reference = parameters.reference;
+    let status = parameters.status;
+    let message = MessageBody {
```
It's the body of the SQS message. It's for the output queue so I've called it `output_message_body`

MancunianSam commented on lambda/src/lib.rs:
```diff
@@ -0,0 +1,202 @@
+    let output_tar_path = process_package(output_path, &input_file_path)?;
+    let file_name = output_tar_path
+        .file_name()
+        .and_then(|oss| oss.to_str())
```
OsString. I've called it `file_name_as_os_string`

MancunianSam commented on anonymise/src/lib.rs:
```diff
@@ -0,0 +1,328 @@
+}
+
+/// # Anonymise the contact fields and update the checksum
+fn update_json_file(
```
Again, we've got anonymise everywhere. This is updating the json file so I think it's ok.

MancunianSam commented on lambda/src/lib.rs:
```diff
@@ -0,0 +1,202 @@
+/// # Processes the SQS message.
+///
+/// This will download the file specified in the message body, anonymise it, upload it to S3 and send the message on to the output queue.
+pub async fn process_record(
```
1. The first argument is of type SqsMessage so it feels like overkill to name it.
2. The method is split up into functions. There's `upload` `download` and `process_package`. I tried moving the code to send the sqs message out but it ends up taking about 5 parameters which is worse I think.

MancunianSam commented on tests/integration_test.rs:
```diff
@@ -1,71 +1,11 @@
+use testlib::*;
 
 #[test]
 fn creates_a_valid_test_package() -> Result<(), Box<dyn std::error::Error>> {
```
It's calling the script binary directly as if you were running it on the command line so it belongs with the command line package.
